src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -117,3 +117,36 @@ def is_satisfied_by(self, candidate):
         # already implements the prerelease logic, and would have filtered out
         # prerelease candidates if the user does not expect them.
         return self.specifier.contains(candidate.version, prereleases=True)
+
+
+class ConstraintRequirement(Requirement):
```

Comment:
We should add a docstring here, detailing how this works / "flows" through the resolver.

https://github.com/pypa/pip/projects/5#card-37162340

Reply:
How about this?

```python
    """A requirement that represents a constraint.

    This requirement acts just like a SpecifierRequirement, except that
    it doesn't generate any real matches of its own. This is to ensure that
    if the user doesn't actually request that package X gets installed, a
    constraint on X doesn't trigger installation.

    ConstraintRequirement objects generate a singleton "dummy" candidate,
    so that they satisfy the condition that any requirement has at least
    one match - this is a resolvelib condition.
    """
```

Reply:
Sounds great, though we're going in a different direction now. :)

Reply:
I don't think we are. But I suspect you've just not caught up on my various ramblings below yet, as you're doing release management at the moment ;-)